WrightTools/collection/_directory.py

```diff
@@ -0,0 +1,90 @@
+"""Cary."""
+
+
+# --- import --------------------------------------------------------------------------------------
+
+
+import fnmatch
+import queue
+import os
+import posixpath
+
+import numpy as np
+
+from .. import exceptions as wt_exceptions
+from ._collection import Collection
+
+
+# --- define --------------------------------------------------------------------------------------
+
+
+__all__ = ['from_directory']
+
+
+# --- from function -------------------------------------------------------------------------------
+
+
+def from_directory(filepath, from_methods, *, name=None, parent=None, verbose=True):
+    """Create a WrightTools Collection from a directory of source files.
+
+    Parameters
+    ----------
+    filepath: str
+        Path to the directory on the file system
+    from_methods: dict<str, callable>
+        Dictionary which maps patterns (using Unix-like glob wildcard patterns)
+        to functions which take a filepath, plus the keyword arguments
+        ['name', 'parent', and 'verbose'].
+        (e.g. most from_<kind> methods within WrightTools)
+        The value can be `None` which results in that item being ignored.
+        The *first* matching pattern encountered will be used.
+        Therefore, if multiple patterns will match the same file, use and `OrderedDict`.
+        Patterns are matched on the file name level, not using the full path.
+
+    Keyword Arguments
+    -----------------
+    name: str
+        Name to use for the root data object. Default is the directory name.
+    parent: Collection
+        Parent collection to insert the directory structure into. Default is a new
+        collection in temp file.
+    verbose: bool
+        Print information as objects are created. Passed to the functions.
+
+    Examples
+    --------
+    >>> from_dict = {'*.data':wt.data.from_PyCMDS,
+    ...              '*.csv':wt.collections.from_Cary,
+    ...              'unused':None,
+    ...             }
+    >>> col = wt.collection.from_directory('path/to/folder', from_dict)
+    """
+    if name is None:
+        name = os.path.basename(os.path.abspath(filepath))
+
+    if verbose:
+        print('Creating Collection:', name)
+
+    root = Collection(name=name, parent=parent)
+
+    q = queue.Queue()
+
+    for i in os.listdir(filepath):
+        q.put((filepath, i, root))
+
+    while not q.empty():
+        path, fname, parent = q.get()
+        for pattern, func in from_methods.items():
+            if fnmatch.fnmatch(fname, pattern):
+                if func is not None:
+                    func(os.path.join(path, fname), name=os.path.splitext(fname)[0], parent=parent,
```

Comment:
why pass in `name`? shouldn't the from method get to decide what the best data object name is?

I guess using this strategy you are guaranteeing unique names

(non-blocking comment)

Reply:
Failing to pass name causes the following exception:

```
  File "<input>", line 1, in <module>
    col = wt.collection.from_directory('.', {'KENT':None, 'COLORS':None, '*.data
':wt.data.from_PyCMDS, '*.csv':wt.collection.from_Cary, '*.dpt':wt.data.from_Ten
sor27}, verbose=True)
  File "/home/kyle/wright/wrighttools/WrightTools/collection/_directory.py", lin
e 81, in from_directory
    verbose=verbose)
  File "/home/kyle/wright/wrighttools/WrightTools/data/_pycmds.py", line 88, in 
from_PyCMDS
    data.create_variable(name='labtime', values=values)
  File "/home/kyle/wright/wrighttools/WrightTools/data/_data.py", line 525, in c
reate_variable
    id = self.require_dataset(name=name, data=values, shape=shape, dtype=dtype).
id
  File "/usr/lib/python3.6/site-packages/h5py/_hl/group.py", line 134, in requir
e_dataset
    raise TypeError("Shapes do not match (existing %s vs new %s)" % (dset.shape,
 shape))
TypeError: Shapes do not match (existing (51,) vs new (81, 81))
```

I'm not sure right now exactly why, but suspect it does indeed have to do with ensuring unique names.

My goal was to replicate the file structure as closely as possible.

Reply:
My personal opinion is that this method is designed for use in cases such as publication figure generation scripts, where the user may wish to override the 'name' field as stored in a `.data` file for instance. I would rather encourage changing to informative data file names rather than encouraging changing of the contents of the file. I also don't wish to require users to manually rename things after calling this.